`trackTransformationSteps` option in `eo-maven-plugin` reports cryptic parsing errors

Here is what I'm doing in my `pom.xml`:

```
<trackTransformationSteps>true</trackTransformationSteps>
```

During the build, the folder `5-pre-transpile` being generated, however I getting this:

```text
/target
  /1-parse # all properly parsed XMIRs, no errors
  /2-pull
  /4-resolve
  /5-pre-transpile # for each XMIR, we have multiple detailed XMIRs, resulted after XSL decorations, e.g. `00-set-locators.xml`, `01-set-original-names.xml`, etc.
  /5-transpile # the only `org/eolang/error.xmir` is present here
```

```text
Caused by: java.lang.IllegalStateException: XMIR should have '/object/o/@name' attribute
    at org.eolang.parser.OnDefault.lambda$get$0 (OnDefault.java:49)
    at java.util.Optional.orElseThrow (Optional.java:403)
    at org.eolang.parser.OnDefault.get (OnDefault.java:48)
    at org.eolang.maven.MjTranspile.lambda$transpilation$2 (MjTranspile.java:196)
    at org.cactoos.func.StickyFunc.lambda$new$0 (StickyFunc.java:51)
    at org.cactoos.func.StickyBiFunc.apply (StickyBiFunc.java:74)
    at org.cactoos.func.StickyFunc.apply (StickyFunc.java:58)
    at org.eolang.maven.TrSpy.lambda$new$0 (TrSpy.java:43)
    at com.yegor256.xsline.StLambda.apply (StLambda.java:115)
    at com.yegor256.xsline.StAfter.lambda$new$0 (StAfter.java:50)
    at com.yegor256.xsline.StLambda.apply (StLambda.java:115)
    at com.yegor256.xsline.StEnvelope.apply (StEnvelope.java:61)
    at com.yegor256.xsline.Xsline.pass (Xsline.java:85)
    at org.eolang.maven.MjTranspile.lambda$transpilation$3 (MjTranspile.java:200)
    at org.eolang.maven.MjTranspile.lambda$transpiled$1 (MjTranspile.java:171)
    at org.eolang.maven.FpDefault.lambda$new$1 (FpDefault.java:71)
    at org.cactoos.func.UncheckedFunc.lambda$apply$0 (UncheckedFunc.java:37)
    at org.cactoos.scalar.Checked.value (Checked.java:56)
    at org.cactoos.scalar.IoChecked.value (IoChecked.java:44)
    at org.cactoos.scalar.Unchecked.value (Unchecked.java:37)
    at org.cactoos.func.UncheckedFunc.apply (UncheckedFunc.java:38)
    at org.eolang.maven.FpGenerated.apply (FpGenerated.java:33)
    at org.eolang.maven.FpUpdateBoth.lambda$new$0 (FpUpdateBoth.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpFork.lambda$new$1 (FpFork.java:44)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpIfTargetOlder.apply (FpIfTargetOlder.java:18)
    at org.eolang.maven.FpFork.lambda$new$1 (FpFork.java:44)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpFork.lambda$new$1 (FpFork.java:44)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpFork.lambda$new$1 (FpFork.java:44)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpFork.lambda$new$1 (FpFork.java:44)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpExistedSource.lambda$new$0 (FpExistedSource.java:25)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.FpEnvelope.apply (FpEnvelope.java:31)
    at org.eolang.maven.MjTranspile.transpiled (MjTranspile.java:178)
    at org.eolang.maven.MjTranspile.lambda$exec$0 (MjTranspile.java:126)
    at org.eolang.maven.Threaded.lambda$total$1 (Threaded.java:76)
    at org.cactoos.scalar.CallableEnvelope.call (CallableEnvelope.java:34)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:642)
    at java.lang.Thread.run (Thread.java:1583)
```

The same build, with the option set to false, gives me clean build:

```text
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/TrSpy.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/TrSpy.java
@@ -5,7 +5,6 @@
 package org.eolang.maven;
 
 import com.jcabi.log.Logger;
-import com.jcabi.xml.XML;
 import com.yegor256.xsline.Shift;
 import com.yegor256.xsline.StAfter;
 import com.yegor256.xsline.StLambda;

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/TrSpy.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/TrSpy.java
@@ -13,10 +13,14 @@ import com.yegor256.xsline.TrEnvelope;
 import com.yegor256.xsline.TrLambda;
 import com.yegor256.xsline.Train;
 import java.nio.file.Path;
-import org.cactoos.Func;
 
 /**
  * Train that spies.
+ * <p>The directory is fixed for the whole train, because a document halfway
+ * through it no longer carries the name of the object it came from: after
+ * {@code classes.xsl} an XMIR holding two objects has two {@code class}
+ * elements and no {@code /object/o/@name} at all, and deriving the directory
+ * from such a document breaks the build, see #4370.</p>
  * @since 0.23
  */
 final class TrSpy extends TrEnvelope {
@@ -26,7 +30,7 @@ final class TrSpy extends TrEnvelope {
      * @param train Original one
      * @param dir The dir to save
      */
-    TrSpy(final Train<Shift> train, final Func<XML, Path> dir) {
+    TrSpy(final Train<Shift> train, final Path dir) {
         super(
             new TrLambda(
                 train,
@@ -38,7 +42,7 @@ final class TrSpy extends TrEnvelope {
                             final String log = shift.uid().replaceAll("[^A-Za-z0-9]", "-");
                             new Saved(
                                 xml.toString(),
-                                dir.apply(xml).resolve(String.format("%02d-%s.xml", pos, log))
+                                dir.resolve(String.format("%02d-%s.xml", pos, log))
                             ).value();
                             if (Logger.isDebugEnabled(TrSpy.class)) {
                                 Logger.debug(

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/Transpilation.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/Transpilation.java
@@ -4,7 +4,6 @@
  */
 package org.eolang.maven;
 
-import com.github.lombrozo.xnav.Xnav;
 import com.jcabi.log.Logger;
 import com.jcabi.xml.XML;
 import com.yegor256.xsline.Shift;
@@ -22,9 +21,6 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.function.Function;
 import java.util.stream.Stream;
-import org.cactoos.func.StickyFunc;
-import org.eolang.parser.OnDefault;
-import org.eolang.parser.OnDetailed;
 import org.eolang.parser.TrFull;
 
 /**
@@ -180,23 +176,15 @@ final class Transpilation {
      * Build XSL transformation function for a source file.
      * If transformation steps are tracked - creates a new {@link Xsline}
      * for every XMIR in purpose of thread safety.
-     * @param source Path to source XMIR
+     * @param name Name of the object the source XMIR holds
      * @return XSL transformation function
      */
-    Function<XML, XML> forSource(final Path source) {
+    Function<XML, XML> forSource(final String name) {
         final Train<Shift> measured = this.measured(this.train());
         final Function<XML, XML> func;
         if (this.tracking.steps()) {
-            func = xml -> new Xsline(
-                new TrSpy(
-                    measured,
-                    new StickyFunc<>(
-                        doc -> new Place(
-                            new OnDetailed(new OnDefault(new Xnav(doc.inner())), source).get()
-                        ).make(this.target.resolve(Transpiling.PRE), "")
-                    )
-                )
-            ).pass(xml);
+            final Path dir = new Place(name).make(this.target.resolve(Transpiling.PRE), "");
+            func = xml -> new Xsline(new TrSpy(measured, dir)).pass(xml);
         } else {
             func = new Xsline(measured)::pass;
         }

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/Transpiling.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/Transpiling.java
@@ -181,12 +181,11 @@ final class Transpiling implements Step {
         final Path source = tojo.xmir();
         final XML xmir = new XMLDocument(source);
         final Path base = this.targetDir.resolve(Transpiling.DIR);
-        final Path target = new Place(
-            new OnDetailed(new OnDefault(new Xnav(xmir.inner())), source).get()
-        ).make(base, MjAssemble.XMIR);
+        final String name = new OnDetailed(new OnDefault(new Xnav(xmir.inner())), source).get();
+        final Path target = new Place(name).make(base, MjAssemble.XMIR);
         final Supplier<String> hsh = new TojoHash(tojo);
         final AtomicBoolean rewrite = new AtomicBoolean(false);
-        final Function<XML, XML> transform = this.train.forSource(source);
+        final Function<XML, XML> transform = this.train.forSource(name);
         final Path cdir = this.cacheDir.resolve(Transpiling.CACHE);
         final Path tail = base.relativize(target);
         if (this.cacheEnabled) {

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/MjTranspileTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/MjTranspileTest.java
@@ -85,18 +85,8 @@ final class MjTranspileTest {
     void tracksStepsOfProgramWithTwoObjects(@Mktmp final Path temp) throws IOException {
         MatcherAssert.assertThat(
             "the first tracked step of a program holding two objects did not leave its XMIR in the pre-transpile directory",
-            new FakeMaven(temp).withProgram(
-                String.join(
-                    System.lineSeparator(),
-                    "+package examples",
-                    "",
-                    "# First.",
-                    "[] > x",
-                    "",
-                    "# Second.",
-                    "[] > y"
-                )
-                ).with("trackTransformationSteps", true)
+            new FakeMaven(temp).withProgram(MjTranspileTest.pair())
+                .with("trackTransformationSteps", true)
                 .execute(MjParse.class)
                 .execute(MjTranspile.class)
                 .result(),
@@ -111,18 +101,8 @@ final class MjTranspileTest {
         throws IOException {
         MatcherAssert.assertThat(
             "the second object of a tracked program did not reach the generated Java",
-            new FakeMaven(temp).withProgram(
-                String.join(
-                    System.lineSeparator(),
-                    "+package examples",
-                    "",
-                    "# First.",
-                    "[] > x",
-                    "",
-                    "# Second.",
-                    "[] > y"
-                )
-                ).with("trackTransformationSteps", true)
+            new FakeMaven(temp).withProgram(MjTranspileTest.pair())
+                .with("trackTransformationSteps", true)
                 .execute(MjParse.class)
                 .execute(MjTranspile.class)
                 .result(),
@@ -560,6 +540,23 @@ final class MjTranspileTest {
     private static String instantiating() {
         return String.format(
             "+package foo.x%n%n[] > main%n  [] > inner%n    42 > @%n  42.plus > @%n    []"
+        );
+    }
+
+    /**
+     * The EO program holding two top-level objects.
+     * @return Source code of the program
+     */
+    private static String pair() {
+        return String.join(
+            System.lineSeparator(),
+            "+package examples",
+            "",
+            "# First.",
+            "[] > x",
+            "",
+            "# Second.",
+            "[] > y"
         );
     }
 

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/MjTranspileTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/MjTranspileTest.java
@@ -107,6 +107,30 @@ final class MjTranspileTest {
     }
 
     @Test
+    void transpilesSecondObjectOfProgramWhileTrackingSteps(@Mktmp final Path temp)
+        throws IOException {
+        MatcherAssert.assertThat(
+            "the second object of a tracked program did not reach the generated Java",
+            new FakeMaven(temp).withProgram(
+                String.join(
+                    System.lineSeparator(),
+                    "+package examples",
+                    "",
+                    "# First.",
+                    "[] > x",
+                    "",
+                    "# Second.",
+                    "[] > y"
+                )
+                ).with("trackTransformationSteps", true)
+                .execute(MjParse.class)
+                .execute(MjTranspile.class)
+                .result(),
+            Matchers.hasKey("target/generated/org/eolang/EO_examples/EOy.java")
+        );
+    }
+
+    @Test
     void wrapsObjectsIntoPhCoverageWhenTrackingEnabled(@Mktmp final Path temp) throws Exception {
         MatcherAssert.assertThat(
             "the generated Java must wrap located objects into PhCoverage when coverageTracking is on",

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/MjTranspileTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/MjTranspileTest.java
@@ -82,6 +82,31 @@ final class MjTranspileTest {
     }
 
     @Test
+    void tracksStepsOfProgramWithTwoObjects(@Mktmp final Path temp) throws IOException {
+        MatcherAssert.assertThat(
+            "the first tracked step of a program holding two objects did not leave its XMIR in the pre-transpile directory",
+            new FakeMaven(temp).withProgram(
+                String.join(
+                    System.lineSeparator(),
+                    "+package examples",
+                    "",
+                    "# First.",
+                    "[] > x",
+                    "",
+                    "# Second.",
+                    "[] > y"
+                )
+                ).with("trackTransformationSteps", true)
+                .execute(MjParse.class)
+                .execute(MjTranspile.class)
+                .result(),
+            Matchers.hasKey(
+                String.format("target/%s/examples/x/00-set-locators.xml", Transpiling.PRE)
+            )
+        );
+    }
+
+    @Test
     void wrapsObjectsIntoPhCoverageWhenTrackingEnabled(@Mktmp final Path temp) throws Exception {
         MatcherAssert.assertThat(
             "the generated Java must wrap located objects into PhCoverage when coverageTracking is on",

